### PR TITLE
test: hot restart IPv6

### PR DESCRIPTION
**Pull request by hennna**
This PR parameterizes IPv4/IPv6 testing for the hot restart integration test. 

See #1036 , #953 , and #1049 .

### CODE REVIEW

File: tools/socket_passing.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/python2.7
```
Comment (htuch) [inline tools/socket_passing.py]: I don't think everyone will be running precisely Python 2.7 or have Python in this location. Why the change?

File: test/integration/hotrestart_test.sh [hunk 2]
```diff
@@ -7,19 +7,40 @@ set -e
...
+    sed -e "s#{{ ip_loopback_address }}#[::1]#" | \
+    sed -e "s#{{ dns_lookup_family }}#v6_only#" | \
+    cat > "${HOT_RESTART_JSON_V6}"
+  JSON_TEST_ARRAY+=("${HOT_RESTART_JSON_V6}")
+fi
+
+TEST_INDEX=0
+for HOT_RESTART_JSON in "${JSON_TEST_ARRAY[@]}"
+do
 # Now start the real server, hot restart it twice, and shut it all down as a basic hot restart
```
Comment (htuch) [inline test/integration/hotrestart_test.sh]: This should all be indented.

File: test/integration/hotrestart_test.sh [hunk 3]
```diff
@@ -7,19 +7,40 @@ set -e
...
-  sed -e "s#{{ test_rundir }}#$TEST_RUNDIR#" | \
-  sed -e "s#{{ ip_loopback_address }}#127.0.0.1#" | \
-  sed -e "s#{{ dns_lookup_family }}#v4_only#" | \
-  cat > "${HOT_RESTART_JSON}"
-
+JSON_TEST_ARRAY=()
+
+# Parameterize IPv4 and IPv6 testing.
+if [[ "${ENVOY_IP_TEST_VERSIONS}" == "v4only" ]] || [[ "${ENVOY_IP_TEST_VERSIONS}" == "all" ]] \
+  || [[ -z "${ENVOY_IP_TEST_VERSIONS}" ]]; then
```
Comment (htuch) [inline test/integration/hotrestart_test.sh]: It would make sense to have this as the first expression to shortcircuit the other testing.

File: tools/socket_passing.py (hunk 1 above)
Comment (hennna) [inline tools/socket_passing.py]: Python 2.3/2.4 doesn't support with open().

Comment (htuch) [inline tools/socket_passing.py]: What about `/usr/bin/env python2.7` then? You can't assume Python is in `/usr/bin`.
